src/stm32_generic/i2c.cpp

```diff
@@ -0,0 +1,548 @@
+          continue;
+        }
+        bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable);
+        m_data_in[0] = data;
+
+        while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+          continue;
+        }
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop);
+        m_data_in[1] = data;
+        m_data_in[2] = data;
+        m_busy = false;
+        break;
+      }
+
+      case 1: {
+        while (bit_extract<i2c_sr1::byte_transfered_finish>(status)) {
+          continue;
+        }
+        m_data_in[0] = data;
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop);
+        m_busy = false;
+        break;
+      }
+
+      case 0: {
+        bit_modify(i2c_reg->cr1).set(i2c_cr1::stop).clear(i2c_cr1::ack_enable);
+        [[maybe_unused]] auto a = data;
+        break;
+      }
+
+      default: {
+        m_data_in[0] = data;
+        m_data_in = m_data_in.subspan(1);
+      }
+    }
+  }
+}
+
+void i2c::handle_i2c_error() noexcept
+{
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  bool soft_reset = false;
+  auto& status = i2c_reg->sr1;
+
+  if (bit_extract<i2c_sr1::bus_error>(status)) {
+    m_status = error_state::io_error;
+    bit_modify(i2c_reg->sr1).clear<i2c_sr1::bus_error>();
+    soft_reset = true;
+  }
+
+  if (bit_extract<i2c_sr1::ack_failure>(status)) {
+    m_status = error_state::no_such_device;
+    bit_modify(i2c_reg->sr1).clear<i2c_sr1::ack_failure>();
+    soft_reset = true;
+  }
+
+  if (bit_extract<i2c_sr1::timeout_error>(status)) {
+    /// we don't support stm32's timeout function, but have an external io
+    /// waiter
+    bit_modify(i2c_reg->sr1).clear(i2c_sr1::timeout_error);
+  }
+
+  if (bit_extract<i2c_sr1::arbitration_lost>(status)) {
+
+    m_status = error_state::arbitration_lost;
+    soft_reset = true;
+  }
+
+  if (soft_reset) {
+    bit_modify(i2c_reg->cr1).set(i2c_cr1::software_reset);
+    bit_modify(i2c_reg->cr1).clear(i2c_cr1::software_reset);
+  }
+}
+void i2c::transaction(hal::byte p_address,
+                      std::span<hal::byte const> p_data_out,
+                      std::span<hal::byte> p_data_in,
+                      hal::function_ref<hal::timeout_function> p_timeout)
+{
+  m_status = error_state::no_error;
+  m_address = p_address;
+  m_data_out = p_data_out;
+  m_data_in = p_data_in;
+  m_busy = true;
+
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  bit_modify(i2c_reg->cr1)
+    .set(i2c_cr1::peripheral_enable)
+    .set(i2c_cr1::ack_enable)
+    .clear(i2c_cr1::stop);
+  bit_modify(i2c_reg->cr1).set(i2c_cr1::start);
+  while (m_busy) {
+    try {
+      p_timeout();
+      m_waiter->wait();
+    } catch (...) {
+      // The expected exception is hal::timed_out, but it could be something
+      // else. Let rethrow the exception so the caller handle it.
+      // A better option here would be to use a std::scope_failure handler.
+      bit_value(i2c_reg->cr1).clear(i2c_cr1::peripheral_enable);
+      m_waiter->resume();
```

Comment:
```suggestion
```

This is the wrong place for resume. Resume should be in the interrupt service routine only, to break the thread out of blocking.